async_simple/Promise.h

```diff
@@ -49,6 +49,9 @@ class Promise {
         _sharedState->attachPromise();
     }
     Promise& operator=(const Promise& other) {
+        if (_sharedState) {
```

Comment:
I prefer `this->~Promise();`. This is more stable if we add other logic to `~Promise()` then.

Reply:
👍🏻，updated

Reply:
Remove copy constructor is also acceptable, I believe passing a promise to another attached promise is not a good choice.
The only reason is std::function requires captured objects should be copyable, but we can use MoveWrapper here.

Reply:
Agreed. This is what Seastar do now: https://github.com/scylladb/seastar/blob/ea6a6820ede9995cbfc6037e430b8d733a088c8f/include/seastar/core/future.hh#L832-L833.

Reply:
ok，I try to remove it again.

Reply:
@RainMark sorry, I took an oversight. Seastar didn't delete the copy assignment. It just defines it out-of-line. https://github.com/scylladb/seastar/blob/ea6a6820ede9995cbfc6037e430b8d733a088c8f/include/seastar/core/future.hh#L973-L979

Reply:
Copy assignment is used in `forward_to`: https://github.com/scylladb/seastar/blob/ea6a6820ede9995cbfc6037e430b8d733a088c8f/include/seastar/core/future.hh#L1721

Reply:
@ChuanqiXu9 your link pointed to move assignment?

Reply:
looks like seastar use noncopyable function to workaround? https://github.com/scylladb/seastar/blob/ea6a6820ede9995cbfc6037e430b8d733a088c8f/include/seastar/util/noncopyable_function.hh#L102

Reply:
Oh, my bad. Thanks for pointing me.

Reply:
> looks like seastar use noncopyable function to workaround? https://github.com/scylladb/seastar/blob/ea6a6820ede9995cbfc6037e430b8d733a088c8f/include/seastar/util/noncopyable_function.hh#L102

From its comment, it looks like our MoveWrapper, is it?

Reply:
@ChuanqiXu9 em, not same, them not use std::function, so no need to wrap the promise

such as the timer's callback: https://github.com/scylladb/seastar/blob/ea6a6820ede9995cbfc6037e430b8d733a088c8f/include/seastar/core/timer.hh#L84

but we already used the std::function in base class interface, so we keep the current usage